### PR TITLE
more SETcc improvements to backend

### PR DESCRIPTION
**Pull request by WalterBright**
Make better use of SETcc instructions.

### CODE REVIEW

File: src/backend/cod4.c [hunk 1]
```diff
@@ -3479,27 +3482,40 @@ code *cdbtst(elem *e, regm_t *pretregs)
...
+            code *nop = CNIL;
+            regm_t save = regcon.immed.mval;
+            code *cg = allocreg(&retregs,&reg,TYint);
+            regcon.immed.mval = save;
+            if ((*pretregs & mPSW) == 0)
+            {
+                cg = cat(cg,getregs(retregs));
+                cg = genregs(cg,0x19,reg,reg);              // SBB reg,reg
+                cg = gen2(cg,0xF7,modregrmx(3,3,reg));      // NEG reg
+            }
```
Comment (MartinNowak) [inline src/backend/cod4.c]: The latency of SBB+NEG is 3 vs. 2 for SETcc+MOVZX, so you might want to use SETcc here as well though it requires 4 bytes for the code.

File: src/backend/cod4.c [hunk 2]
```diff
@@ -3479,27 +3482,40 @@ code *cdbtst(elem *e, regm_t *pretregs)
...
-            nop = gennop(nop);
-            cg = genjmp(cg,JC,FLcode,(block *) nop);    // Jtrue nop
-                                                        // MOV reg,0
-            movregconst(cg,reg,0,8);
-            regcon.immed.mval &= ~mask[reg];
+            code *nop = CNIL;
+            regm_t save = regcon.immed.mval;
+            code *cg = allocreg(&retregs,&reg,TYint);
+            regcon.immed.mval = save;
+            if ((*pretregs & mPSW) == 0)
```
Comment (MartinNowak) [inline src/backend/cod4.c]: What does `mPSW` stand for?

Comment (yebblies) [inline src/backend/cod4.c]: Processor status word aka flags register

**Comment by WalterBright** [comment]
While I wait impatiently for ddmd (hint, hint) I decided to do some tuneups to the back end. If you notice any code sequences where clang/gcc do a better job than dmd, let me know.

File: src/backend/cod4.c (hunk 1 above)
Comment (MartinNowak) [inline src/backend/cod4.c]: That's only beneficent for loops.

**Comment by MartinNowak** [comment]
Auto-merge toggled on

**Pull request merged**